crypto.cpp

```diff
@@ -123,126 +133,484 @@ OvpnCryptoUninitAlgHandles(_In_ BCRYPT_ALG_HANDLE aesAlgHandle, BCRYPT_ALG_HANDL
+{
+    OvpnPktidRecv* recv = NULL;
+
+    if (epoch == 0 || keySlot->Decrypt.Epoch == epoch) {
+        recv = &keySlot->PktidRecv;
+    }
+    else if (epoch == keySlot->RetiringEpochDataReceiveKey.Epoch) {
+        recv = &keySlot->PktidRecvRetiring;
+    }
+    else {
+        /* We have an epoch that is neither current or old recv key but
+         * is authenticated, ie we need to move to a new current recv key */
+        LOG_INFO("Received data packet with new epoch. Updating receive key", TraceLoggingValue(epoch, "epoch"));
+        OvpnCryptoEpochReplaceUpdateRecvKey(keySlot, epoch, opts);
+        recv = &keySlot->PktidRecv;
+    }
+
+    return OvpnPktidRecvVerify(recv, packet_id_net);
+}
+
 static
 NTSTATUS
-OvpnCryptoAEADDoWork(BOOLEAN encrypt, OvpnCryptoKeySlot* keySlot, UCHAR *bufIn, SIZE_T len, UCHAR* bufOut, INT32 cryptoOptions)
+OvpnCryptoAEADDoWork(BOOLEAN encrypt, OvpnCryptoKeySlot* keySlot, UCHAR *bufIn, SIZE_T len, UCHAR* bufOut, OvpnCryptoOptions* opts)
 {
     /*
     AEAD Nonce :
 
      [Packet ID] [HMAC keying material]
-     [4/8 bytes] [8/4 bytes           ]
+     [4 bytes  ] [4 bytes             ]
      [AEAD nonce total : 12 bytes     ]
 
     TLS wire protocol :
 
-    Packet ID is 8 bytes long with CRYPTO_OPTIONS_64BIT_PKTID.
-
      [DATA_V2 opcode] [Packet ID] [AEAD Auth tag] [ciphertext]
-     [4 bytes       ] [4/8 bytes] [16 bytes     ]
+     [4 bytes       ] [4 bytes  ] [16 bytes     ]
      [AEAD additional data(AD)  ]
 
-    With CRYPTO_OPTIONS_AEAD_TAG_END AEAD Auth tag is placed after ciphertext:
+    New data format, with epoch keys and 64bit packet id:
+
+     struct aead_packet {
+        int opcode:5;
+        int key_id:3;
+        int peer_id:24;
+        uint64_t packet_id;
+        uint8_t* encrypted_payload;
+        uint8_t[16] authentication_tag;
+     }
+
+     struct packet_id {
+        uint epoch:16;
+        uint epoch_counter:48;
+     }
+
+     authenticated_data = opcode| key_id | peer_id | packet_id
 
-     [DATA_V2 opcode] [Packet ID] [ciphertext] [AEAD Auth tag]
-     [4 bytes       ] [4/8 bytes]              [16 bytes     ]
-     [AEAD additional data(AD)  ]
     */
 
     NTSTATUS status = STATUS_SUCCESS;
 
-    BOOLEAN pktId64bit = cryptoOptions & CRYPTO_OPTIONS_64BIT_PKTID;
-
-    SIZE_T cryptoOverhead = OVPN_DATA_V2_LEN + AEAD_AUTH_TAG_LEN + (pktId64bit ? 8 : 4);
+    BOOLEAN authTagEnd = opts->UseEpoch;
+    ULONG pktidLen = opts->UseEpoch ? 8 : 4;
+    ULONG cryptoOverhead = OVPN_DATA_V2_LEN + AEAD_AUTH_TAG_LEN + pktidLen;
 
     if (len < cryptoOverhead) {
         LOG_WARN("Packet too short", TraceLoggingValue(len, "len"));
         return STATUS_DATA_ERROR;
     }
 
+    // we prepended buf with crypto overhead
+    len -= cryptoOverhead;
+
+    OvpnCryptoKeyContext* decryptKey = &keySlot->Decrypt;
+    UINT64 packet_id = 0;
+    UINT16 rx_epoch = 0;
+
     UCHAR nonce[12];
     if (encrypt) {
```

Comment:
Maybe split this method into encrypt and decrypt? This has become quite large by now.

Reply:
Yes, will do. In fact there are already separate Encrypt and Decrypt functions, which both call DoWork. I agree that the single function now makes less sense.